0005-kinesis-proxy-json.md

```diff
@@ -0,0 +1,306 @@
+- missing metadata: the messages do not include the date they were sent, or
+  the time the messages were received by SocketProxy 
+
+To address these issues, we'll create a KinesisProxy to write the OCS messages
+(along with additional metadata) as JSON CloudEvent messages to an AWS
+Kinesis stream. The partition key will be based on the source `host:port`
+address.
+
+# Guide-level explanation
+
+AWS Kinesis "makes it easy to collect, process, and analyze real-time,
+streaming data so you can get timely insights and react quickly to new
+information." Unlike a message queue (SQS or RabbitMQ), messages written to
+AWS Kinesis can be repeatedly read by multiple clients.
+
+Kinesis consists of named *streams*, and each stream consists of one or more
+*shards*. *Records* are read/written to the stream, and distributed between
+the shards based on a *partition key*. *Producers* write records to the
+stream, and are read by zero or more *consumers*. When written, each record
+receives a *sequence number*: this number is always increasing and uniquely
+identifies the record within the shard. It also orders the messages: messages
+within a shard have a defined order, but not messages received by two
+different shards.
+
+In this case, a new KinesisProxy will be the producer. For each message
+received from the OCS, it will transform that message into a CloudEvent (see
+[RFC 4](https://github.com/mbta/technology-docs/pull/4)), encode the event as
+JSON, and put a record in the "ctd-raw-ocs-messages" stream. The
+"ctd-raw-ocs-messages" stream will be configured with one (1) shard. In order
+that train movement messages are processed in the correct order, the
+partition key will be based on the source `host:port` values. This ensures
+that all messages from a given connection are kept in order.
+
+The four (4) RTR instances will be consumers of the stream. Instead of
+expecting messages on TCP sockets, they will read the CloudEvent JSON payload
+from the stream, parse out the raw message, and send it through the existing
+pipeline.
+
+A [JSON Schema](https://json-schema.org/) document will be written in the
+`mbta/schemas` repository, with the name of the CloudEvent message.
+
+# Reference-level explanation
+
+<!-- This is the technical portion of the RFC. Explain the design in sufficient detail that:
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.  The section should return to the
+examples given in the previous section, and explain more fully how the
+detailed proposal makes those examples work. -->
+
+Currently, SocketProxy has a Receiver GenServer responsible for each incoming
+connection from OCS. Receiver then spawns a Forwarder GenServer to send the
+data to each of the destinations (currently, the four RTR instances). After
+this RFC, a new KinesisProxy which is responsible for
+writing records to a configured Kinesis stream:
```

Comment:
I'm guessing this is the new Elixir app which will replace SocketProxy. You might not have recognized it as an app name by Paul because it doesn't have some cutesy name like... `Tricycle`, with many levels of meaning. Probably need to open a new RFC to properly bikeshed the name. :)

Reply:
Now I kind of want to call it Tricycle :)